src/main/scala/com/microsoft/hyperspace/actions/RefreshDeleteAction.scala

```diff
@@ -0,0 +1,92 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.actions
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.functions._
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.util.ResolverUtils
+
+/**
+ * Refresh index by removing index entries for deleted source data files.
+ * If some original data file(s) are removed, in order to update index,
+ * user can trigger an index refresh action during which:
+ * 1. Deleted source data files are detected;
+ * 2. Index records' lineage is leveraged to remove any index entry coming from deleted files.
+ *
+ * @param spark SparkSession
+ * @param logManager Index LogManager for index being refreshed
+ * @param dataManager Index DataManager for index being refreshed
+ */
+class RefreshDeleteAction(
+    spark: SparkSession,
+    logManager: IndexLogManager,
+    dataManager: IndexDataManager)
+    extends RefreshActionBase(spark, logManager, dataManager) {
+
+  /**
+   * For an index with lineage, find all the source data files which have been deleted,
+   * and use index records' lineage to mark and remove index entries which belong to
+   * deleted source data files as those entries are no longer valid.
+   */
+  final override def op(): Unit = {
+    val indexDF = spark.read.parquet(previousIndexLogEntry.content.files.map(_.toString): _*)
+
+    ResolverUtils
+      .resolve(spark, IndexConstants.DATA_FILE_NAME_COLUMN, indexDF.schema.fieldNames) match {
+      case Some(_) =>
+        val refreshDF =
+          indexDF.filter(
+            !col(s"${IndexConstants.DATA_FILE_NAME_COLUMN}").isin(getDeletedFiles: _*))
+
+        refreshDF.write.saveWithBuckets(
+          refreshDF,
+          indexDataPath.toString,
+          logEntry.asInstanceOf[IndexLogEntry].numBuckets,
+          indexConfig.indexedColumns)
+
+      case None =>
+        throw HyperspaceException(s"Refresh delete is only supported on an index with lineage.")
+    }
+  }
+
+  /**
+   * Compare list of source files from previous index log entry to list of current data files
+   * and detect deleted source files.
+   *
+   * @return list of full paths of deleted source files.
+   */
+  private def getDeletedFiles: Seq[String] = {
+    val rels = previousIndexLogEntry.relations
+    // Currently we only support to create an index on a LogicalRelation.
+    assert(rels.size == 1)
+
+    val originalFiles = rels.head.data.properties.content.files.map(_.toString)
+    val currentFiles = rels.head.rootPaths.flatMap { p =>
+      Content
+        .fromDirectory(path = new Path(p))
```

Comment:
`throwIfNotExists` should be true?

Reply:
I dont think so. Data file path does not exist is not necessarily unexpected. Imagine a corner case where user deletes all the source data files and calls incremental refresh (instead of deleting and vacuuming index). Then this function should return all the original source data files as deleted files.

Reply:
good point. can you add a test case if it's not covered?

Reply:
while adding suggested test case, I realized due to current impl the code fails before even reaching this point (before calculating the diff between files list).
Consider below cases:
**Case 1. User deletes source data folder completely (i.e. both folder and its content are gone)**
Refresh Action fails as the very first call to `event()` API, which happens at the beginning of any action, makes a chain of calls as
`override def logEntry: LogEntry = getIndexLogEntry(..)` -> `protected lazy val df = {..}`
and `df` call fails since it does a `spark.read.load` action to read a DataFrame from a path that does not exist. (i.e. `DataFrameReader` fails with Path does not exist exception).

**Case 2. User deletes all the source data files under source data folder but keeps the folder**
Simiar to the above case, `event()` call in this case ends up calling `getIndexLogEntry(..)` from `CreateActionBase` and that fails as the DataFrame created from an empty folder which has no real relevant data file does not have a valid Hyperspace signature: The call fails with `HyperspaceException("Invalid plan for creating an index.")`.

So such a scenario (user deleting all source data files and calling refresh) fails in its very initial steps and before getting to actual refresh action touching/fixing an index. As a result, I added your original suggestion (`throwIfNotExists = true`) to be semantically consistent with the expectations, although we should never really hit it. 

But one side question here is: Are we fine with above way of failure or we need to change the way failure surfaces to the user? (CC @rapoth )

Reply:
Ok, I need some more context to contribute to this discussion. :) @pirz Can you please update both your cases with the appropriate error message that will be thrown?

Reply:
@rapoth I updated above comment with the error user sees for each case. 

1. In Case 1 (Deleting whole source data folder), the error comes from Spark's DataFrameReader.
2. In Case 2 (Deleting all data files under source data folder but keeping the folder), the error is captured by Hyperspace code and a Hyperspace Exception is thrown.